src/host/readDataDirect.cpp

```diff
@@ -88,105 +78,59 @@ bool DirectReadData::Notify(const WaitTerminationReason TerminationReason,
-        {
-            *pReplyStatus = STATUS_INTEGER_OVERFLOW;
-            return retVal;
-        }
-        size_t amountToRead;
-        if (FAILED(SizeTSub(_eventReadCount, amountAlreadyRead, &amountToRead)))
-        {
-            *pReplyStatus = STATUS_INTEGER_OVERFLOW;
-            return retVal;
-        }
-
-        *pReplyStatus = _pInputBuffer->Read(readEvents,
-                                            amountToRead,
-                                            false,
-                                            false,
-                                            fIsUnicode,
-                                            false);
-
-        if (*pReplyStatus == CONSOLE_STATUS_WAIT)
-        {
-            retVal = false;
-        }
+        *pReplyStatus = STATUS_INTEGER_OVERFLOW;
+        return true;
     }
 
-    if (*pReplyStatus != CONSOLE_STATUS_WAIT)
+    *pReplyStatus = _pInputBuffer->Read(_outEvents,
+                                        amountToRead,
+                                        false,
+                                        false,
+                                        fIsUnicode,
+                                        false);
+
+    if (*pReplyStatus == CONSOLE_STATUS_WAIT)
     {
-        // split key events to oem chars if necessary
-        if (*pReplyStatus == STATUS_SUCCESS && !fIsUnicode)
-        {
-            try
-            {
-                SplitToOem(readEvents);
-            }
-            CATCH_LOG();
-        }
-
-        // combine partial and whole events
-        while (!_partialEvents.empty())
-        {
-            readEvents.push_front(std::move(_partialEvents.back()));
-            _partialEvents.pop_back();
-        }
-
-        // move read events to out storage
-        for (size_t i = 0; i < _eventReadCount; ++i)
-        {
-            if (readEvents.empty())
-            {
-                break;
-            }
-            _outEvents.push_back(std::move(readEvents.front()));
-            readEvents.pop_front();
-        }
-
-        // store partial event if necessary
-        if (!readEvents.empty())
-        {
-            _pInputBuffer->StoreReadPartialByteSequence(std::move(readEvents.front()));
-            readEvents.pop_front();
-            FAIL_FAST_IF(!(readEvents.empty()));
-        }
-
-        // move events to pOutputData
-        const auto pOutputDeque = reinterpret_cast<std::deque<std::unique_ptr<IInputEvent>>* const>(pOutputData);
-        *pNumBytes = _outEvents.size() * sizeof(INPUT_RECORD);
-        pOutputDeque->swap(_outEvents);
+        return false;
     }
-    return retVal;
+
+    // move events to pOutputData
+    const auto pOutputDeque = static_cast<std::deque<std::unique_ptr<IInputEvent>>* const>(pOutputData);
+    *pNumBytes = _outEvents.size() * sizeof(INPUT_RECORD);
+    *pOutputDeque = std::move(_outEvents);
+
+    return true;
+}
+catch (...)
+{
+    // There doesn't seem to be a way to just get the NT exception code with WIL.

```

Comment:
you might benefit from `NT_CATCH_RETURN()`! It's in `wil/nt_result_macros.h`

Reply:
That doesn't quite work, since it uses an out parameter. But I just found `wil::StatusFromCaughtException()`!